### ISSUE TITLE
About the  practical implementation of object memory

### ISSUE THREAD

**Issue by firrice** [hkchengrex/Cutie#129]
Hi, Cutie is a great work, however there was a problem annoying me which was about the actual implementation of pooling masks within object memory. In the raw paper, pooling masks was defined as below:
![image](https://github.com/user-attachments/assets/b4e50eee-eaae-462e-80aa-ad997b5895ca)
But in the code, it turns out that:
![image](https://github.com/user-attachments/assets/a2b11934-c4ca-4dc7-963e-a306f5e976f9)
It seems like that the two zero conditions was actually replaced by multiplying with repeated_masks which was defined:
![image](https://github.com/user-attachments/assets/a10b994c-d856-4924-b503-261c0e95fbd3)
I wander if i got it right, look forward to your reply,thanks!

**Comment by hkchengrex** [comment]
Hi, thank you for the question!

`masks` and `inv_masks` contain binary values (1 or 0) and they act as masks to the `logits.sigmoid()` term. You can think of the computation as
1. We compute condition (3) for every feature
2. If the feature satisfies condition (1), indicated by `masks`, we set it to zero by masking
3. If the feature satisfies condition (2), indicated by `inv_masks`, we set it to zero by masking

Effectively the result is the same. Feel free to ask if this isn't clear!

**Comment by firrice** [comment]
> Hi, thank you for the question!
> 
> `masks` and `inv_masks` contain binary values (1 or 0) and they act as masks to the `logits.sigmoid()` term. You can think of the computation as
> 
> 1. We compute condition (3) for every feature
> 2. If the feature satisfies condition (1), indicated by `masks`, we set it to zero by masking
> 3. If the feature satisfies condition (2), indicated by `inv_masks`, we set it to zero by masking
> 
> Effectively the result is the same. Feel free to ask if this isn't clear!

Thanks for your so quick reply! I have also considered that maybe "masks and inv_masks contain binary values (1 or 0) " , and in this way, the final result can be equal to the three conditions in paper. However,  I have found that masks was not binary tensor through debugging step by step. Further more, in order to validate it, i have also extracted the top-100 values from masks(its shape was [batchsize, object_num, 30, 54, 1])，the corresponding visualization code and top-100 elements were depicted as below:
![image](https://github.com/user-attachments/assets/57252db2-620c-4f4b-a727-6589b757a0cd)
According to values in the yellow box beyond to the png above, it seems like that the masks tensor was indeed non-binary, and thus results can not be same. Am i undertstanding it right?

**Comment by firrice** [comment]
> > Hi, thank you for the question!
> > `masks` and `inv_masks` contain binary values (1 or 0) and they act as masks to the `logits.sigmoid()` term. You can think of the computation as
> > 
> > 1. We compute condition (3) for every feature
> > 2. If the feature satisfies condition (1), indicated by `masks`, we set it to zero by masking
> > 3. If the feature satisfies condition (2), indicated by `inv_masks`, we set it to zero by masking
> > 
> > Effectively the result is the same. Feel free to ask if this isn't clear!
> 
> Thanks for your so quick reply! I have also considered that maybe "masks and inv_masks contain binary values (1 or 0) " , and in this way, the final result can be equal to the three conditions in paper. However, I have found that masks was not binary tensor through debugging step by step. Further more, in order to validate it, i have also extracted the top-100 values from masks(its shape was [batchsize, object_num, 30, 54, 1])，the corresponding visualization code and top-100 elements were depicted as below: ![image](https://private-user-images.githubusercontent.com/26590498/393056801-57252db2-620c-4f4b-a727-6589b757a0cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0NTI1NTYsIm5iZiI6MTczMzQ1MjI1NiwicGF0aCI6Ii8yNjU5MDQ5OC8zOTMwNTY4MDEtNTcyNTJkYjItNjIwYy00ZjRiLWE3MjctNjU4OWI3NTdhMGNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA2VDAyMzA1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjNDgyMDg1MGJjZmNhYjY1OTFlMmQ3MjJhNGJkZjFjZDcxMTA2MDdlYzRiNjAxZGNmODFmMGZjMzA5OGEyMDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MVjDGAPbwDosyjLSA1rSmQV9dAefhH_IWqSiakilJdc) According to values in the yellow box beyond to the png above, it seems like that the masks tensor was indeed non-binary, and thus results can not be same. Am i undertstanding it right?

Additional note: The excuted entry script was "scripting_demo.py"

**Comment by hkchengrex** [comment]
Ah, it seems like the resize operation is introducing some non-binary values. They are probably inconsequential and the output would be a smooth interpolation between the conditions so I don't expect that to be a problem. Thank you for pointing this out. I think you are right.

**Comment by firrice** [comment]
> Ah, it seems like the resize operation is introducing some non-binary values. They are probably inconsequential and the output would be a smooth interpolation between the conditions so I don't expect that to be a problem. Thank you for pointing this out. I think you are right.

Thanks for so patient reply! i have got it clearly :)
